File: lana/src/main/scala/com/financialforce/lana/display/OpenFileInPackage.scala
/*
 * Copyright (c) 2020 FinancialForce.com, inc. All rights reserved.
 */
package com.financialforce.lana.display

import com.financialforce.lana.Context
import com.financialforce.lana.runtime.vscode._

import scala.scalajs.concurrent.JSExecutionContext.Implicits.queue

object OpenFileInPackage {

  def openFileForSymbol(wsPath: String, context: Context, name: String, line: Option[Int]): Unit = {
    context
      .findSymbol(wsPath, name)
      .foreach(path => {
        val uri = Uri.file(path.toString)
        workspace
          .openTextDocument(uri)
          .toFuture
          .flatMap(window.showTextDocument(_).toFuture.map { editor =>
            line.foreach(line => {
              val position = new Position(line, character = 0)
              editor.revealRange(new Range(position, position))
            })
          })
      })
  }
}

File: ext/src/display/OpenFileInPackage.ts
/*
 * Copyright (c) 2020 FinancialForce.com, inc. All rights reserved.
 */

import { Position, Range, Uri, window, workspace } from "vscode";
import { Context } from "../Context";

export class OpenFileInPackage {
  static openFileForSymbol(
    wsPath: string,
    context: Context,
    name: string,
    line?: number
  ): void {
    const path = context.findSymbol(wsPath, name);
    if (path) {
      const uri = Uri.file(path);
      workspace.openTextDocument(uri).then((td) => {
        window.showTextDocument(td).then((editor) => {
          if (line) {
            const position = new Position(line, 0);
            editor.revealRange(new Range(position, position));
          }
        });
      });
    }
  }
}
